### PR TITLE
bluespace miner

### PR DESCRIPTION
**Pull request by jjpark-kb**
## About The Pull Request
adds the bluespace miner, a machine that is capable of  "mining" materials through the power of bluespace! (read: it creates materials).
There are a few restrictions:
1) It needs to be in a room with a temp less than room temp
2) It needs a specific pressure range of the room
2a) It needs to be in a room with pressure that is above the normal room
2b) It needs to be in a room with pressure that is less than twice the normal room
3) It cannot be directly near another bluespace miner

## How This Contributes To The Skyrat Roleplay Experience
material scarcity really doesn't matter, especially in Skyrat. Perhaps having more materials will allow us more freedoms.

## Changelog

:cl:
add: added the bluespace miner
/:cl:

### CODE REVIEW

File: code/modules/projectiles/projectile/energy/nuclear_particle.dm [hunk 1]
```diff
@@ -3,7 +3,7 @@
 	name = "nuclear particle"
 	icon_state = "nuclear_particle"
 	pass_flags = PASSTABLE | PASSGLASS | PASSGRILLE
-	flag = TOX
+	flag = ENERGY
```
Comment (tf-4) [inline code/modules/projectiles/projectile/energy/nuclear_particle.dm]: skyrat edit

Comment (jjpark-kb) [inline code/modules/projectiles/projectile/energy/nuclear_particle.dm]:
```suggestion
	flag = ENERGY //SKYRAT EDIT: was TOX
```

Comment (jjpark-kb) [reply code/modules/projectiles/projectile/energy/nuclear_particle.dm]: resolved.

**Comment by thestubborn** [comment]
gmod money printers lets GOOOOOOOOOOOO

**Comment by softcerv** [comment]
> gmod money printers lets GOOOOOOOOOOOO

having money printers back would lead to a lot of fun set ups

**Comment by Stalkeros2** [comment]
> having money printers back would lead to a lot of fun set ups

:thinking:

**Comment by Iamgoofball** [comment]
:-1:
Infinite materials is not good. We explicitly removed the ORM resource quadrupling for this purpose.

**Comment by francinum** [comment]
Why the fuck do I agree with goofball for once.

Material scarcity isn't a problem for several reasons

This isn't helping, it's *making the problem worse*

**Comment by jjpark-kb** [comment]
> Why the fuck do I agree with goofball for once.
> 
> Material scarcity isn't a problem for several reasons
> 
> This isn't helping, it's _making the problem worse_

You aren't explaining whats the problem and how this is making the problem worse. Also, who are you?

**Comment by ghost** [comment]
Honestly, I dislike the idea of materials being gotten without interacting with Cargo or at least off-station for them;
Having science and atmospherics make materials (without sciences already-existing xenobio material duplication) just feels almost... out of place, to me at least.

A better way to do this might be the drills that Shiptest uses, where you can scan for ore veins and deploy a drill, which if upgraded will teleport the materials to an output; this way, there'd be setup for each material, the veins could optionally be made limited, and it would still require to at least get the materials from off-station(even if its just from asteroids in space)

And yes, I know, "WYCI", but I'm not complaining about this idea, just saying there might be a better way than 'bluespace'
~~plus it'd pair well with Overmap~~

**Comment by jjpark-kb** [comment]
It can only be purchased, which means it has to go through Cargo. Ideally, this would be bought by Cargo (because it is expensive) and would be a collaboration between Cargo and Atmos.

Drills are in Overmap, so I won't be adding that.

File: modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+	name = "bluespace miner"
+	desc = "Through the power of bluespace, it is capable of producing materials."
+	icon = 'modular_skyrat/modules/bluespace_miner/icons/bluespace_miner.dmi'
+	icon_state = "miner"
+
+	density = TRUE
+	circuit = /obj/item/circuitboard/machine/bluespace_miner
+	idle_power_usage = 200
+
+	///the temperature of the co2 produced per successful process (its really 100)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm]: What unit are you using? Kelvins or Celcius?

File: modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+/obj/machinery/bluespace_miner/examine(mob/user)
+	. = ..()
+	var/turf/src_turf = get_turf(src)
+	var/datum/gas_mixture/environment = src_turf.return_air()
+	if(environment.temperature >= T20C)
+		. += span_warning("[src] is in a suboptimal environment: TOO HOT TEMPERATURE!")
+	if(environment.return_pressure() <= ONE_ATMOSPHERE)
+		. += span_warning("[src] is in a suboptimal environment: TOO LITTLE PRESSURE!")
+	if(environment.return_pressure() >= (ONE_ATMOSPHERE * 2))
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm]: It shouldn't have both at once, so might as well reduce the checks it does.
```suggestion
	else if(environment.return_pressure() >= (ONE_ATMOSPHERE * 2))
```

File: modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	if(machine_stat & (NOPOWER|BROKEN))
+		return
+	add_overlay("miner_on")
+
+/obj/machinery/bluespace_miner/examine(mob/user)
+	. = ..()
+	var/turf/src_turf = get_turf(src)
+	var/datum/gas_mixture/environment = src_turf.return_air()
+	if(environment.temperature >= T20C)
+		. += span_warning("[src] is in a suboptimal environment: TOO HOT TEMPERATURE!")
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm]:
```suggestion
		. += span_warning("[src] is in a suboptimal environment: TEMPERATURE TOO HIGH!")
```

File: modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+	if(environment.return_pressure() <= ONE_ATMOSPHERE)
+		. += span_warning("[src] is in a suboptimal environment: TOO LITTLE PRESSURE!")
+	if(environment.return_pressure() >= (ONE_ATMOSPHERE * 2))
+		. += span_warning("[src] is in a suboptimal environment: TOO MUCH PRESSURE!")
+	for(var/obj/machinery/bluespace_miner/bs_miner in range(1, src))
+		if(bs_miner != src)
+			. += span_warning("[src] is in a suboptimal environment: TOO CLOSE TO ANOTHER BLUESPACE MINER")
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm:60]:
```suggestion
			. += span_warning("[src] is in a suboptimal environment: TOO CLOSE TO ANOTHER BLUESPACE MINER!")
```

Comment (GoldenAlpharex) [reply modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm:60]: Might want to make all the stuff in all caps, also bolded, so it's clearer what changes.

File: modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+		return FALSE
+	//add amount_produced degrees to the temperature
+	var/datum/gas_mixture/merger = new
+	merger.assert_gas(/datum/gas/carbon_dioxide)
+	merger.gases[/datum/gas/carbon_dioxide][MOLES] = MOLES_CELLSTANDARD
+	merger.temperature = (T20C + gas_temp)
+	src_turf.assume_air(merger)
+	return TRUE
+
+/obj/machinery/bluespace_miner/proc/spawn_mats()
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm:99]: Documentation please :)

File: modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+		. += span_warning("[src] is in a suboptimal environment: TOO HOT TEMPERATURE!")
+	if(environment.return_pressure() <= ONE_ATMOSPHERE)
+		. += span_warning("[src] is in a suboptimal environment: TOO LITTLE PRESSURE!")
+	if(environment.return_pressure() >= (ONE_ATMOSPHERE * 2))
+		. += span_warning("[src] is in a suboptimal environment: TOO MUCH PRESSURE!")
+	for(var/obj/machinery/bluespace_miner/bs_miner in range(1, src))
+		if(bs_miner != src)
+			. += span_warning("[src] is in a suboptimal environment: TOO CLOSE TO ANOTHER BLUESPACE MINER")
+
+/obj/machinery/bluespace_miner/proc/check_factors()
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm:63]: Documentation please :)

File: modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+		/obj/item/stack/ore/bluespace_crystal/refined = 1,
+		/obj/item/stock_parts/matter_bin = 2,
+		/obj/item/stock_parts/micro_laser = 2,
+		/obj/item/stock_parts/manipulator = 2)
+	needs_anchored = TRUE
+
+/datum/supply_pack/misc/bluespace_miner
+	name = "Bluespace Miner"
+	desc = "Nanotrasen has revolutionized the procuring of materials with bluespace-- featuring the Bluespace Miner!"
+	cost = CARGO_CRATE_VALUE * 2500 // 500,000, one of the most expensive buys (besides the bicycle, rip)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm]: I'm still not convinced that this is the right way to go about it, but we'll see.

I certainly don't want us to buff cargo so they make that much money every round, just so they can afford a bluespace miner, though.

**Review by GoldenAlpharex — CHANGES_REQUESTED** [review]
Also, if we do end up with this merged, we probably would need a better sprite for it.

File: modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm (hunk 2 above)
Comment (jjpark-kb) [reply modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm]: its kelvin

File: modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm (hunk 7 above)
Comment (jjpark-kb) [reply modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm:63]: done

File: modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm (hunk 6 above)
Comment (jjpark-kb) [reply modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm:99]: done

File: modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm (hunk 5 above)
Comment (jjpark-kb) [reply modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm:60]: done

File: modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm (hunk 4 above)
Comment (jjpark-kb) [reply modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm]: done

File: modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm (hunk 3 above)
Comment (jjpark-kb) [reply modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm]: done

File: modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm (hunk 8 above)
Comment (jjpark-kb) [reply modular_skyrat/modules/bluespace_miner/code/bluespace_miner.dm]: Yea, we can always change how to obtain it depending on how it performs.

Oh yea no, I do not want to buff cargo getting more money just to get these.

**Review by GoldenAlpharex — APPROVED** [review]
Code checks out, let's test it out.

**Comment by Zenitheevee** [comment]
I'll put it here to be a bit more formal:
If you _Can_: Add a description to the circuitboard that states it requires atmos and temperature control (Maybe that the circuitry is sensitive to overheating, and needs a certain pressure?), and maybe mention that it'll drop the mats, So people can plan accordingly.

First test, Despite the messy execution, Went off well enough i have faith in it, and had i not known it needed atmos control beforehand i wouldn't have built it in the Tarkon Mixing chamber. (Which may have been a bit metagamey for me but i did want to atleast make sure it functioned)

**Comment by jjpark-kb** [comment]
Good suggestion, I'll do so.

**Comment by Gandalf2k15** [comment]
We will only see this as endgame tech and honestly if people want to waste resources on it as a T4 experimental, let them.

**Pull request merged**